### PR TITLE
Fix issue #670 (Destructuring struct returned from func call bug)

### LINKED ISSUE #670: Bug when Numerically Operating Casted Variable (Destructured from Struct)
# Description

## Aim

Variables bound from structs, regardless of via the period (`.`) or destructuring by pattern, should be castable and numerically operable.

## Expected behavior

Both of the following excerpts should successfully prove when running `nargo prove p`.

- Period (`.`) Example:

```rust
struct Animal {
    legs: Field,
    eyes: u8,
}

fn main() -> pub Field {
    let dog = Animal { legs: 4, eyes: 2 };
    let six = dog.legs + dog.eyes as Field; // <-- This line
    six
}
```

- Destructuring Example:

```rust
struct Animal {
    legs: Field,
    eyes: u8,
}

fn main() -> pub Field {
    let Animal { legs, eyes } = get_dog();
    let six = legs + eyes as Field; // <-- Versus this line
    six
}

fn get_dog() -> Animal {
    let dog = Animal { legs: 4, eyes: 2 };
    dog
}
```

## Bug

Proving the Period (`.`) Example is successful:

```
% nargo prove p
ACIR opcodes generated : 5
...
Ok
```

Proving the Destructuring Example gives:

```
% nargo prove p
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Unsigned(8)`,
 right: `NativeField`', crates/noirc_evaluator/src/ssa/node.rs:746:17
```

# Environment

For `nargo` users

- _noir-lang/noir_ commit cloned: [00b64a2d1f0838fad9df1aba5b15dd293cd80138](https://github.com/noir-lang/noir/commit/00b64a2d1f0838fad9df1aba5b15dd293cd80138) (Compiled from source)

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #670

# Description

Reference issue for details of the bug.

## Summary of changes

In code_gen, when we are returning a struct from a function we have a flattened vector of fields. Struct fields are stored as an ordered list, thus when we map the NodeIds returned from a func call in the IRGenerator we are associating NodeIds with an ordered list of struct fields. When we construct an `ExtractTupleMember` expression in the monomorphisation, we unpack a struct pattern with several let statements for each struct field in the way they were declared by the user. Thus, if the user does not deconstruct their struct in alphanumerical order when the struct is returned from a func call this could lead to an incorrect object being accessed. 

I altered the patterns iterator used in the struct case of `unpack_pattern` to use a BTreeMap. This will match the order of `fields` in the `StructType` which is being unpacked. When calling `unpack_tuple_pattern` we enumerate over the patterns iterator and use the current index at each iteration to construct an `ExtractTupleField` expression. This has not changed, except now that the struct pattern iterator uses a BTreeMap we are accessing the correct index.  

## Test additions / changes

I added a regression test to our existing `struct` test that uses the snippet from the linked issue this PR is solving. 

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

### CODE REVIEW

File: crates/noirc_frontend/src/monomorphisation/mod.rs [hunk 1]
```diff
@@ -396,11 +396,13 @@ impl Monomorphiser {
...
-                let patterns = patterns.into_iter().map(|(ident, pattern)| {
+                // We map each pattern to its respective field in a BTreeMap
+                // Fields in struct types are ordered, and doing this map guarantees we extract the correct field index
+                let patterns_map = btree_map(patterns, |(ident, pattern)| {
                     let typ = fields[&ident.0.contents].clone();
-                    (pattern, typ)
+                    (ident.0.contents, (pattern, typ))
                 });
-                self.unpack_tuple_pattern(value, patterns)
+                self.unpack_tuple_pattern(value, patterns_map.values().cloned())
```
Comment (jfecher) [inline crates/noirc_frontend/src/monomorphisation/mod.rs:405]: You're looking for `.into_values()`, no need to clone when you already own the BTreeMap

**Review by jfecher — CHANGES_REQUESTED** [review]
Nice, straightforward solution

File: crates/noirc_frontend/src/monomorphisation/mod.rs (hunk 1 above)
Comment (vezenovm) [reply crates/noirc_frontend/src/monomorphisation/mod.rs]: Ah I didn't see method. Thank you, done

File: crates/nargo/tests/test_data/struct/src/main.nr [hunk 2]
```diff
@@ -59,5 +68,10 @@ fn main(x: Field, y: Field) {
     constrain a_bool == true;
     constrain struct_from_tuple.my_int == 5;
     constrain struct_from_tuple.my_nest.a == 0;
+
+    // Regression test for issue #670
+    let Animal { legs, eyes } = get_dog();
+    let six = legs + eyes as Field; // <-- Versus this line
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply crates/nargo/tests/test_data/struct/src/main.nr]: Removed

**Review by jfecher — APPROVED** [review]
👍

**Comment by vezenovm** [comment]
> 👍

Will merge when CI finishes

**Pull request merged**